59.md

```diff
@@ -0,0 +1,292 @@
+
+# Encrypting Payloads
+
+Encryption is done following NIP-44 on the JSON-encoded event. Place the the encryption payload in the `.content` of the wrapper event (either a `seal` or a `gift wrap`).
+
+# Other Considerations
+
+If a `rumor` is intended for more than one party, or if the author wants to retain an encrypted copy, a single `rumor` may be wrapped multiple times, with each wrapper addressed to a different recipient.
+
+The canonical `created_at` time belongs to the `rumor`. All other timestamps SHOULD be tweaked to thwart time-analysis attacks. Note that some relays don't serve events dated in the future, so all timestamps SHOULD be in the past.
+
+Relays may choose not to store gift wrapped events due to them not being publicly useful. Clients MAY choose to attach a certain amount of proof-of-work to the wrapper event per NIP-13 in a bid to demonstrate that the event is not spam or a denial-of-service attack.
+
+To protect recipient metadata, relays SHOULD only serve kind 1059 events intended for the marked recipient. When possible, clients should only send wrapped events to read relays for the recipient that implement AUTH, and refuses to serve wrapped events to non-recipients.
+
+# An Example
+
+Let's send a wrapped `kind 1` message between two parties asking "Are you going to the party tonight?"
+
+- Author private key: `0beebd062ec8735f4243466049d7747ef5d6594ee838de147f8aab842b15e273`
+- Recipient private key: `e108399bd8424357a710b606ae0c13166d853d327e47a6e5e038197346bdbf45`
+- Ephemeral wrapper key: `4f02eac59266002db5801adc5270700ca69d5b8f761d8732fab2fbf233c90cbd`
+
+## 1. Create an event
+
+Create a `kind 1` event with the message, the receivers, and any other tags you want, signed by the author. Do not sign the event.
+
+```json
+{
+  "kind": 1,
+  "created_at": 1691518405,
+  "pubkey": "611df01bfcf85c26ae65453b772d8f1dfd25c264621c0277e1fc1518686faef9",
+  "content": "Are you going to the party tonight?",
+  "tags": [["p", "166bf3765ebd1fc55decfe395beff2ea3b2a4e0a8946e7eb578512b555737c99"]],
+  "id": "90297600a3b91c41a8a044993efb7f0b568e5944020bfee9c50ad7dae0d8c729"
+}
+```
+
+## 2. Seal the rumor
+
+Encrypt the JSON-encoded `rumor` with a conversation key derived using the **Sender**'s private key and the recipient's public key. JSON-encode the resulting object and place it in the `content` field of a `kind 13` `seal` event. Sign it with the **Sender**'s key.
+
+```json
+{
+  "content": "{\"ciphertext\":\"f5Lfq6a9laLW9PWuYatTTnrGOSHcGJUPnb5OQ/gEEVjtDnrOzDB9VFKdUkoImEMiO6Ddu+HYn0R3hUtqgTRppOjs4zNFTLNAP/3hkN3hAaKjvUzBVYVmU1xrBS749kUhCt8v/8l/tBeen9BYB4s1luAUqdS10h8jwM+NqE8aYrv9sEU9AaLuErbI1hwq0ZUEGl5unfxHpv+QjINbrPYs1x6bGUU7D416Lp0gV6fr/wD/JLkceb/quAE2UkgzVRS4gauWL8/PyaPz3332h+vH/tMWOQd2no4sDG3ftXPO8y0zXs5p/NvcPa2hIXsOsNOPzOfCFvBk7Rfv9JjwCN0K/LztDHZOzgryELAjcm+mDWPeYBw+PYWfwayFkDbbLCsQJ1KLYRBDS7BEMBOLdd1v1vF4J196tGaN\",\"nonce\":\"s6DBwU6gq9h14+9hRwnts1qZ6BdyPgHn\",\"v\":1}",
```

Comment:
is there a reason these fields are as json in the content instead of just tags?

Reply:
The payload is defined by NIP 44 (#715), so it can't be split across tags. @vitorpamplona wanted it to be json. I personally prefer @paulmillr's original version that is CSV oriented, but Paul has also suggested doing `base64.encode(JSON.stringify({cyphertext, nonce, v}))`. It's just bikeshedding at this point.

Reply:
@vitorpamplona what is the problem with large tags?

Reply:
as someone writing a relay and database, all I can say is anytime I have to parse json it's a pita. tags are in a decoded form already.

Reply:
> @vitorpamplona what is the problem with large tags?

Indexing of large tags is bad. A Relational Implementation of Nostr will not have a choice to remove certain tags before indexing. I would generally avoid breaking their implementation if we can

Reply:
relays only index single-char tags

Reply:
You shouldn't have to parse this payload on a relay unless you're decrypting the payload, it's intended to be opaque.